deep_quoridor/src/agents/alphazero/multiprocess_evaluator.py

```diff
@@ -0,0 +1,223 @@
+        all_games_inputs_array = np.stack([input_array] + [NNEvaluator.game_to_input_array(g) for g in all_games[1:]])
+        all_games_action_masks = np.stack([g.get_action_mask() for g in all_games])
+
+        values_array, policies_array = self.evaluate_arrays(all_games_inputs_array, all_games_action_masks)
+
+        for row_i in range(len(values_array)):
+            # Sanity checks
+            assert np.all(policies_array[row_i] >= 0), "Policy contains negative probabilities"
+            assert np.abs(np.sum(policies_array[row_i]) - 1) < 1e-6, "Policy does not sum to 1"
+            assert np.isfinite(values_array[row_i]), "Policy or value is non-finite"
+
+            # Update our local cache
+            cache_key = all_games_inputs_array[row_i].tobytes()
+            self._cache[cache_key] = (values_array[row_i], policies_array[row_i], game.copy())
+
+        evaluation_end_time = time.time()
+        self._evaluation_log.append(EvaluationInfo(evaluation_start_time, evaluation_end_time, 1, False))
+
+        value = values_array[0]
+        policy = policies_array[0]
+        if game.is_rotated():
+            policy = policy[self._action_mapping_rotated_to_original]
+
+        return value, policy
+
+    def evaluate_arrays(self, input_arrays: np.ndarray, action_mask_arrays: np.ndarray):
+        self._request_queue.put((input_arrays, action_mask_arrays, self._client_id))
+        values_array, policies_array = self._result_queue.get()
+        return values_array, policies_array
+
+    def rotate_policy_from_original(self, policy: np.ndarray):
+        return policy[self._action_mapping_original_to_rotated]
+
+    def rotate_if_needed_to_point_downwards(self, game: Quoridor):
+        return NNEvaluator.rotate_if_needed_to_point_downwards(game)
+
+    def game_to_input_array(self, game):
+        return NNEvaluator.game_to_input_array(game)
+
+    def clear_cache(self):
+        self._cache.clear()
+
+    def get_statistics(self):
+        return compute_statistics(self._evaluation_log)
+
+
+class EvaluatorServer(threading.Thread):
+    def __init__(
+        self,
+        evaluator: NNEvaluator,
+        input_queue: mp.Queue,
+        output_queues: list[mp.Queue],
+        max_log_len=int(1e8),
+    ):
+        super().__init__()
+        self._evaluator = evaluator
+        self._cache = {}
+        self._input_queue = input_queue
+        self._output_queues = output_queues
+        self._max_l = max_log_len
+        self._evaluation_log = deque(maxlen=max_log_len)
+        self._shutdown = False
+
+    def run(self):
+        while not self._shutdown:
+            try:
+                inputs_array, action_masks_array, client_id = self._input_queue.get(timeout=0.1)
+            except queue.Empty:
+                continue
+
+            values_array, policies_array = self._evaluate_arrays(inputs_array, action_masks_array)
+            self._send_result_to_client(client_id, values_array, policies_array)
```

Comment:
One thing that we can consider in the future is to batch together requests from multiple workers.
I.e. we would loop getting an element from the queue until we hit a timeout and accumulate the inputs.  Then we would evaluate everything together, and use the cache to reply to each client what they expect

Reply:
Yes! I had this implemented earlier but it was hard to set the timeout so that it didn't delay the worker processes too much, and so I removed it. I think it's worth adding back in. @muralx suggested trying it without the timeout - you grab as many items are in the queue, stack them together, and evaluate them at once. This should help if a few requests get backed up in the queue while we're evaluating the previous request.